engine/src/main/java/com/datatorrent/stram/RecoverableRpcProxy.java

```diff
@@ -68,98 +68,111 @@
   private static final int RPC_TIMEOUT_DEFAULT = 5000;
 
   private final Configuration conf;
-  private final String appPath;
   private StreamingContainerUmbilicalProtocol umbilical;
   private String lastConnectURI;
-  private long lastCompletedCallTms;
-  private long retryTimeoutMillis = Long.getLong(RETRY_TIMEOUT, RETRY_TIMEOUT_DEFAULT);
-  private long retryDelayMillis = Long.getLong(RETRY_DELAY, RETRY_DELAY_DEFAULT);
-  private int rpcTimeout = Integer.getInteger(RPC_TIMEOUT, RPC_TIMEOUT_DEFAULT);
-
-  public RecoverableRpcProxy(String appPath, Configuration conf) throws IOException
+  private long retryTimeoutMillis;
+  private long retryDelayMillis;
+  private int rpcTimeout;
+  private final UserGroupInformation currentUser;
+  private final SocketFactory defaultSocketFactory;
+  private final FSRecoveryHandler fsRecoveryHandler;
+
+  public RecoverableRpcProxy(String appPath, Configuration conf)
   {
     this.conf = conf;
-    this.appPath = appPath;
-    connect();
+    try {
+      currentUser = UserGroupInformation.getCurrentUser();
+      defaultSocketFactory = NetUtils.getDefaultSocketFactory(conf);
+      fsRecoveryHandler = new FSRecoveryHandler(appPath, conf);
+      connect(0);
+    } catch (IOException e) {
+      LOG.error("Fail to create RecoverableRpcProxy", e);
+      throw new RuntimeException(e);
+    }
   }
 
-  private void connect() throws IOException
+  private long connect(long timeMillis) throws IOException
```

Comment:
timeMillis is the return value of the connect() in case connect URL was not changed. It is reset on line 130 in the case of updated URL.